src/main/java/com/mongodb/spark/sql/connector/schema/InferSchema.java

```diff
@@ -392,6 +446,52 @@ private static DataType dataTypeCheckStructTypeToMapType(
     return dataType;
   }
 
+  /**
+   * Recursive merge that prefers the rhs datatype value over the lhs. For use with partial schema hints.
+   *
+   * @param lhs the left hand side datatype
+   * @param rhs the right hand side datatype
+   * @return the merged original and preferred data type
+   */
+  @VisibleForTesting
+  static DataType rhsPreferredMerge(final DataType lhs, final DataType rhs) {
+
+    // If they are both structs process the fields and merge.
+    if (lhs instanceof StructType && rhs instanceof StructType) {
+      Map<String, StructField> lhsMap = Arrays.stream(((StructType) lhs).fields())
+          .collect(Collectors.toMap(StructField::name, f -> f));
+      Map<String, StructField> rhsMap = Arrays.stream(((StructType) rhs).fields())
+          .collect(Collectors.toMap(StructField::name, f -> f));
+
+      Map<String, StructField> newFields = new HashMap<>();
+
+      lhsMap.forEach((k, v) -> {
+        if (rhsMap.containsKey(k)) {
+          newFields.put(
+              k,
+              new StructField(
+                  v.name(),
+                  rhsPreferredMerge(v.dataType(), rhsMap.get(k).dataType()),
+                  v.nullable(),
+                  v.metadata()));
+        } else {
+          newFields.put(k, v);
+        }
+      });
+
+      rhsMap.forEach((k, v) -> {
+        if (!newFields.containsKey(k)) {
+          newFields.put(k, v);
+        }
+      });
+
+      return new StructType(newFields.values().toArray(new StructField[0]));
```

Comment:
```suggestion
      // Iterate over rhs fields to merge or add.
      for (StructField rhsField : ((StructType) rhs).fields()) {
        String fieldName = rhsField.name();
        StructField lhsField = lhsMap.get(fieldName);

       // Merge field if exists in lhs; else, add from rhs.
        StructField newField = lhsField == null ? rhsField : new StructField(
                fieldName,
                rhsPreferredMerge(lhsField.dataType(), rhsField.dataType()),
                rhsField.nullable(),
                rhsField.metadata());

        lhsMap.put(fieldName, newField);
      }

      return new StructType(lhsMap.values().toArray(new StructField[0]));
```

It seems we might streamline the code by eliminating the creation of additional `rhsMap` and `newFields` maps. Specifically:

```
Map<String, StructField> rhsMap = Arrays.stream(((StructType) rhs).fields())
    .collect(Collectors.toMap(StructField::name, f -> f));
Map<String, StructField> newFields = new HashMap<>();
```

We could consider directly merging the `rhs` fields into the `lhs` map with overrides, since `lhs` is local and not shared elsewhere. Since `lhsMap` is local and not shared elsewhere, this would allow us to simply return it.

Reply:
Done - renamed to clarify